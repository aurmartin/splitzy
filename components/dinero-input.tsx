import dinero, { Dinero } from "dinero.js";
import React from "react";
import FloatInput from "./float-input";
import { getMinorUnitAmount } from "@/lib/currency";

const DineroInput = (props: {
  value: Dinero;
  onChange: (value: Dinero) => void;
}) => {
  const { value, onChange: onChangeProp } = props;

  const onChange = React.useCallback(
    (numberValue: number) => {
      const currency = value.getCurrency();

      onChangeProp(
        dinero({
          amount: getMinorUnitAmount(numberValue, currency),
          currency,
        }),
      );
    },
    [onChangeProp, value],
  );

  return <FloatInput value={value.toUnit()} onChange={onChange} />;
};

export default DineroInput;
