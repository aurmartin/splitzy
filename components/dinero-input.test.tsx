import { render, screen, userEvent } from "@testing-library/react-native";
import dinero, { type Dinero } from "dinero.js";
import * as React from "react";

import { DineroInput } from "./dinero-input";

describe("DineroInput", () => {
  it("renders correctly", () => {
    const mockOnChange = jest.fn();
    const value = dinero({ amount: 1000, currency: "USD" }); // $10.00

    render(<DineroInput value={value} onChange={mockOnChange} />);
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("converts float input to dinero value correctly", async () => {
    const user = userEvent.setup();

    const mockOnChange = jest.fn((value: Dinero) => value.getAmount());
    const value = dinero({ amount: 1000, currency: "USD" });

    render(
      <DineroInput label="test-input" value={value} onChange={mockOnChange} />,
    );

    const input = screen.getByLabelText("test-input");

    await user.clear(input);
    await user.type(input, "25.50");

    expect(mockOnChange).toHaveLastReturnedWith(2550);
  });
});
