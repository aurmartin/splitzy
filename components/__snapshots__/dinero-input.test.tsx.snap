// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DineroInput renders correctly 1`] = `
<TextInput
  keyboardType="numeric"
  onChangeText={[Function]}
  placeholder="0.00"
  placeholderTextColor="#666"
  style={
    [
      {
        "backgroundColor": "#fff",
        "borderColor": "hsl(0 0% 70%)",
        "borderRadius": 4,
        "borderWidth": 1,
        "height": 45,
        "padding": 10,
      },
      undefined,
    ]
  }
  testID="float-input"
  value="10"
/>
`;
